feat: implement true vesting accounts with clawback

JimLarson opened the pull request:
## Description

Closes: https://github.com/Agoric/agoric-sdk/issues/4085

Implements "true" vesting accounts which are subject to clawback.

True vesting accounts have a lockup schedule and a vesting schedule. Any tokens may be staked, but cannot be withdrawn unless they are both unlocked and vested. At any time, the account which funded the true vesting account may command a "clawback" of the unvested amount. The clawback may result in the transfer of bonded or unbonding tokens.

Staking rewards are encumbered with vesting in proportion to the vested/unvested ratio of the staked tokens.

This PR contains:

- proto definitions for the new account type and messages
- updates the vesting app module and message handler to include a staking keeper
- CLI for account creation and clawback commands
- message handler methods for the new messages
- utilities for merging schedules in various ways
- implementation of new account type with clawback and reward methods
-  hook in distribution keeper for post-reward action on certain accounts
- staking keeper operations to transfer staked and unbonding amounts

The last is the most critical to review, as it mutates core staking data structures.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [X] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [X] provided a link to the relevant issue or specification
- [X] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [PENDING] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [X] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

JimLarson committed:
feat: operations to transfer (un-, re-) delegated tokens

TODO: tests

JimLarson committed:
feat: proto and service definitions for true vesting accounts

The docs mention an obsoleted field - not sure how to fix.

JimLarson committed:
feat: implementation of clawback for true vesting accounts

JimLarson committed:
feat: reward distribution, clawback update of delegation bookkeeping

JimLarson committed:
chore: drop unneeded field

JimLarson committed:
feat: call smart reward handler from distribution module

JimLarson committed:
feat: message types for true vesting accounts

JimLarson committed:
feat: message serving for true vesting accounts

JimLarson committed:
feat: CLI for true vesting commands

michaelfig commented on docs/core/proto-docs.md:
```diff
@@ -578,10 +578,15 @@
+    - [TrueVestingAccount](#cosmos.vesting.v1beta1.TrueVestingAccount)
   
 - [cosmos/vesting/v1beta1/tx.proto](#cosmos/vesting/v1beta1/tx.proto)
+    - [MsgClawback](#cosmos.vesting.v1beta1.MsgClawback)
+    - [MsgClawbackResponse](#cosmos.vesting.v1beta1.MsgClawbackResponse)
```
So, for bikeshedding, I'd suggest something that clearly pairs the account type and clawback messages.  I think "true vesting" needs to go, though, since it seems to imply that this implementation is the best one. 
Some suggestions:

- `ClawbackVestingAccount`, `MsgClawbackVesting`, `MsgCreateClawbackVestingAccount`
- `RevocableVestingAccount`, `MsgRevokeVesting`, and `MsgCreateRevocableVestingAccount`

I should not be the final authority on this.  Please ensure Dean and Zaki make the decision as to the names.

JimLarson replied:
I'll check.

michaelfig commented on proto/cosmos/vesting/v1beta1/tx.proto:
```diff
@@ -46,6 +51,36 @@ message MsgCreatePeriodicVestingAccount {
+// MsgCreateTrueVestingAccountResponse defines the MsgCreateTrueVestingAccount response type.
+message MsgCreateTrueVestingAccountResponse {}
+
+// MsgClaback defines a message that removes unvested (in the true vesting sense) tokens
```
```suggestion
// MsgClawback defines a message that removes unvested (in the true vesting sense) tokens
```

JimLarson replied:
Done.

michaelfig commented on x/auth/vesting/client/cli/tx.go:
```diff
@@ -95,16 +101,44 @@ type InputPeriod struct {
 }
 
+// readScheduleFile reads the file at path and unmarshals it to get the schedule.
+// Returrns start time, periods, and error.
```
```suggestion
// Returns start time, periods, and error.
```

JimLarson replied:
Done.

michaelfig reviewed (commented):
Just posting now so that you have time to act on the bikeshed.  Still reviewing the rest of the PR.

michaelfig commented on x/auth/vesting/msg_server.go:
```diff
@@ -256,5 +185,173 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+		vestingCoins = lockupCoins
+	}
+
+	if !vestingCoins.IsEqual(lockupCoins) { // XXX IsEqual can crash
```
```suggestion
	if !vestingCoins.IsEqual(lockupCoins) { // XXX IsEqual can panic
```

JimLarson replied:
Done.

michaelfig commented on x/auth/vesting/msg_server.go:
```diff
@@ -256,5 +185,173 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+		if !ok {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrNotSupported, "account %s must be a true vesting account", msg.ToAddress)
+		}
+		if msg.FromAddress != pva.FunderAddress {
```
This is probably fine for now.  Unfortunately, though, it allows a funder to squat on well-known addresses and lock out other funders.

JimLarson replied:
I didn't know address scarcity was a problem. If so, then there's a bigger problem of a user racing to create an account at that address with an undesired type, e.g `PermanentLockedAccount`. We could conceivably track vesting grants from multiple funders, but then reward distribution would get really ugly. I hear you say "probably fine for now", so I'll go with that.

michaelfig replied:
This turns out not to be much of a problem, so we're good.

michaelfig commented on x/auth/vesting/msg_server.go:
```diff
@@ -256,5 +185,173 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+	acc := ak.GetAccount(ctx, to)
+
+	if acc != nil {
+		pva, ok := acc.(*types.TrueVestingAccount)
```
Please rename `pva` to `va` here and below, so that we can pivot on the account names without leaving behind cruft from older names:
```suggestion
		va, ok := acc.(*types.TrueVestingAccount)
```

JimLarson replied:
Done.

michaelfig commented on x/auth/vesting/msg_server.go:
```diff
@@ -256,5 +185,173 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+}
+
+// Clawback removes the unvested amount from a TrueVestingAccount.
+// The destination defaults to the funder address, but
```
Incomplete thought?

JimLarson replied:
Completed.

michaelfig commented on x/auth/vesting/msg_server.go:
```diff
@@ -256,5 +185,173 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+	}
+
+	acc := ak.GetAccount(ctx, addr)
+	tva, ok := acc.(*types.TrueVestingAccount)
```
Again, please just `va` instead of `tva`.

JimLarson replied:
Done.

michaelfig commented on x/auth/vesting/types/period.go:
```diff
@@ -26,3 +27,257 @@ func (vp Periods) String() string {
+	return
+}
+
+// AlighSchedules rewrites the first period length to aligh the two arguments to the same start time.
```
```suggestion
// AlignSchedules rewrites the first period length to align the two arguments to the same start time.
```

JimLarson replied:
At least I was consistent. Fixed.

michaelfig commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -611,3 +589,393 @@ func marshalYaml(i interface{}) (interface{}, error) {
+func (tva TrueVestingAccount) Validate() error {
+	// XXX ensure that lockup and vesting schedules both sum to OriginalVesting
+	if tva.GetStartTime() >= tva.GetEndTime() {
+		return errors.New("vesting start-time cannot be before end-time")
```
```suggestion
		return errors.New("vesting start-time must be before end-time")
```

JimLarson replied:
Done.

michaelfig commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -611,3 +589,393 @@ func marshalYaml(i interface{}) (interface{}, error) {
+}
+
+// Clawback transfers unvested tokens in a TrueVestingAccount to dest.
+// Future vesting events are removed. Unstaked tokens
```
Incomplete thought?

JimLarson replied:
I'm lucky to have so many complete thoughts recorded. Fixed.

michaelfig commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -611,3 +589,393 @@ func marshalYaml(i interface{}) (interface{}, error) {
+
+	accPtr := &updatedAcc
+	writeAcc := func() { ak.SetAccount(ctx, accPtr) }
+	writeAcc() // do this now so that unvested tokens are unlocked
```
I think that all store writes will be aborted if there is a stray panic, so this code should be correct, but it may be worthwhile adding that assumption to a comment somewhere in here to make that explicit.

JimLarson replied:
Done.

michaelfig reviewed (commented):
Wow!  Good stuff!

I didn't find anything that looked suspicious, but there are just a few nits I pointed out.

JimLarson committed:
docs: review feedback - fixes to comments and variable names

JimLarson committed:
fix: address to-fix comments by adding checks and fixing rounding

JimLarson commented:
I've addressed all but one to-fix comment, so the code is ready for review.

I'll be concurrently working on the tests. Please request particular tests to cover.

JimLarson committed:
test: schedule utility tests and bugfix

JimLarson committed:
test: vesting account tests, codec registry

JimLarson committed:
fix: properly update vesting delegation bookkeeping

See also https://github.com/Agoric/agoric-sdk/issues/4300

JimLarson committed:
test: handler/msg_server tests

JimLarson committed:
test: tests for transfer of staking state, fixes

zmanian commented on x/staking/keeper/delegation.go:
```diff
@@ -626,6 +627,152 @@ func (k Keeper) Delegate(
 	return newShares, nil
 }
 
+// TransferDelegation changes the ownership of at most the desired number of shares.
+// Returns the actual number of shares transferred. Will also transfer redelegation
+// entries to ensure that all redelegations are matched by sufficient shares.
+// Note that no tokens are transferred to or from any pool or account, since no
+// delegation is actually changing state.
```
My view has been that transfer delegation should be in the upstream staking module. It's useful in a lot of scenarios and I don't think it harms proof of stake security. I think there should be a message type for in the delegation module

fedekunze replied:
Agree, this is useful and should be upstreamed to the SDK

JimLarson replied:
Filed https://github.com/Agoric/agoric-sdk/issues/4315 for a later PR to surface this feature.

zmanian commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -611,3 +589,412 @@ func marshalYaml(i interface{}) (interface{}, error) {
+			if !found {
+				panic("validator not found") // shouldn't happen
+			}
+			wantShares, err := validator.SharesFromTokensTruncated(want)
```
This should handle a slashing event correctly

JimLarson replied:
Do you mean "this should be changed so that it handles slashing" or "it's great that this already handles slashing"?  It looks like validators are never removed if any delegator is still holding shares, but I made the code tolerate a missing validator, reading the delegation as having no value, both here and in similar following code..

zmanian reviewed (approved):
Looked for the obvious edge case like Clawback after slashing and it does the correct things

fedekunze commented on proto/cosmos/vesting/v1beta1/tx.proto:
```diff
@@ -43,8 +48,39 @@ message MsgCreatePeriodicVestingAccount {
   string   to_address                      = 2 [(gogoproto.moretags) = "yaml:\"to_address\""];
   int64    start_time                      = 3 [(gogoproto.moretags) = "yaml:\"start_time\""];
   repeated Period vesting_periods = 4 [(gogoproto.nullable) = false];
+  bool merge = 5;
```
what does `merge` mean, can you add a comment?

JimLarson replied:
It means to merge the described grant into an existing `PeriodicVestingAccount`, or create it if it does not exist.

Added a comment here, and for the like-named field in `MsgCreateTrueVestingAccount`.

fedekunze replied:
what's the use case for merging?

JimLarson replied:
You'll have to ask @dtribble or @rowgraus for the exact justification. We want merging for `TrueVestingAccount` too in order to be able to make staking rewards vest.

fedekunze commented on proto/cosmos/vesting/v1beta1/tx.proto:
```diff
@@ -43,8 +48,39 @@ message MsgCreatePeriodicVestingAccount {
+message MsgCreateTrueVestingAccount {
+  option (gogoproto.equal) = false;
+
+  string from_address = 1 [(gogoproto.moretags) = "yaml:\"from_address\""];
+  string   to_address                      = 2 [(gogoproto.moretags) = "yaml:\"to_address\""];
```
are these the tx signer and vesting addresses? or the vesting and the "admin"?

JimLarson replied:
Since we decided that creating a `ClawbackVestingAccount` (revised name, renaming commit will follow) is a non-privileged operation, we don't want to use the term "admin", so we're describing the one making the grant (and signing it) as the "funder". Added descriptions in all the request message fields to make things more clear.

fedekunze commented on proto/cosmos/vesting/v1beta1/vesting.proto:
```diff
@@ -83,3 +83,21 @@ message PermanentLockedAccount {
+  BaseVestingAccount base_vesting_account = 1 [(gogoproto.embed) = true];
+
+  // funder_address specifies the account which can perform clawback.
+  string funder_address = 2;
```
nit, I'd rename this to `admin`

JimLarson replied:
In this implementation we decided not to make it a privileged operation to create a vesting account, so it would be better to have a name that doesn't suggest a privileged role. Open to suggestions other than "funder", though.

fedekunze replied:
sounds good

fedekunze commented on proto/cosmos/vesting/v1beta1/vesting.proto:
```diff
@@ -83,3 +83,21 @@ message PermanentLockedAccount {
+  // funder_address specifies the account which can perform clawback.
+  string funder_address = 2;
+  int64 start_time = 3 [(gogoproto.moretags) = "yaml:\"start_time\""];
+  repeated Period combined_periods = 4 [(gogoproto.moretags) = "yaml:\"combined_periods\"", (gogoproto.nullable) = false];
```
what are combined periods?

liamsi replied:
They are explained here: https://github.com/Agoric/agoric-sdk/issues/4085#issuecomment-997600471
Probably good to have inline comments for these as well IMO.

But I also don't find it very easy to follow.

liamsi replied:
Can someone answer this:

Let's say someone's tokens are subject to 36 months vesting AND lockup from TGE, released every 6 months.
After month 6, is it:

- Option 1: 1/6th of "subject-to-lockup" tokens become vested, but only 1/6th of "subject-to-lockup" tokens are unlocked, which means that in fact they can only sell 1/36th of those tokens;

OR

- Option 2: the lockup runs concurrently to vesting; e.g. for all tokens regardless of vesting, it is the case that no more than x tokens shall be sold according to some lockup schedule. So after 6 months, 1/6th of tokens will be unlocked and unvested.

Or are both scenarios possible/configurable depending on the schedule of each? If so, how?

JimLarson replied:
I'll add comments here too. Combined periods gives the merge of the unlocking schedule and the vesting schedule, to determine when tokens are actually unencumbered. For instance, if I make a clawback grant of 1000 today (2022-01-17) that vests a 1/4th every 3months, but all of it is locked until Nov 1, then we'd have:

```
Unlocking schedule:
2022-11-01: 1000

Vesting schedule:
2022-04-17: 250
2022-07-17: 250
2022-10-17: 250
2023-01-17: 250

Combined schedule (when tokens are both vested and unlocked):
2022-11-01: 750
2023-01-17: 250
```

Per a note in `vesting_accounts.go`, I'm considering dropping the `combined_periods` field and just querying the two schedules directly. I'd expect the common use case is a vesting schedule with dozens of events, but with a lockup cliff, so this would be just as efficient.

fedekunze replied:
I think we can remove the combined schedule actually 👍

JimLarson replied:
Okay, I'll implement that.

BTW, this PR is fighting against the prevailing terminology in `x/auth/vesting` in that it uses "vesting" only to describe the status associated with clawback, and "lockup" to describe the encumbrance that prevents transfer - i.e. what is called "vesting" elsewhere. Let me know if I need to be even more emphatic about this in the comments.

musalbas replied:
What would you have instead of the combined schedule then? I think it's important to keep it, because Option 1 of the comment above is unlikely to be the understanding of what most people mean by combined vesting and lockup, or the contractual terms.

JimLarson replied:
Oh, we'll have the same behavior, but we just won't precompute the combined schedule.  I.e. to get the unlocked coins at time `t`, instead of computing `ReadSchedule(combined_periods, t)`, we'd compute `min(ReadSchedule(lockup_periods, t), ReadSchedule(vesting_periods, t)`. (Not the exact API calls, of course, but that's the idea.)

fedekunze commented on x/auth/vesting/client/cli/tx.go:
```diff
@@ -166,7 +261,63 @@ func NewMsgCreatePeriodicVestingAccountCmd() *cobra.Command {
+				return err
+			}
 
+			funderString, _ := cmd.Flags().GetString(FlagFunder)
```
shouldn't this be the `from` address?

JimLarson replied:
Ugh - we've got some ugly choices here and I'd appreciate a better suggestion. At clawback time, we're transferring funds **from** the vesting account **to** a destination account, based on a message signed by the funder who was the "from" address when the account was created. So at clawback time, "from" is dangerously ambiguous. My proposed solution here is that at clawback time we say "funder" to avoid ambiguity, but at creation time we say "from" for consistency with the other commands to create vesting accounts.

fedekunze replied:
Tbh I think maintaining `from` is totally OK since this is the CLI (this will remove additional flags) and I'd expect users to use gRPC (which can use the `funder` name)

JimLarson replied:
Turns out to be necessary to get the cli tests to work - at least without more poking into it. Done.

fedekunze commented on x/auth/vesting/msg_server.go:
```diff
@@ -160,5 +185,177 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", msg.DestAddress)
+	}
+
+	acc := ak.GetAccount(ctx, addr)
```
check `acc == nil`

JimLarson replied:
Done.

fedekunze commented on x/auth/vesting/types/expected_keepers.go:
```diff
@@ -2,12 +2,31 @@ package types
 	BlockedAddr(addr sdk.AccAddress) bool
 }
+
+type StakingKeeper interface {
```
what's the purpose of the staking keeper in the context of vesting? I can't find where is this being used

JimLarson replied:
We use the staking keeper to perform clawback: to obtain the balance of bonded and unbonding tokens, and to command the transfer of such tokens.

fedekunze replied:
can you add as a comment on top of the type?

JimLarson replied:
Done.

fedekunze commented on x/auth/vesting/types/msgs.go:
```diff
@@ -13,10 +13,18 @@ const TypeMsgCreateVestingAccount = "msg_create_vesting_account"
 // TypeMsgCreatePeriodicVestingAccount defines the type value for a MsgCreateVestingAccount.
 const TypeMsgCreatePeriodicVestingAccount = "msg_create_periodic_vesting_account"
 
+const TypeMsgCreateTrueVestingAccount = "msg_create_true_vesting_account"
+
+const TypeMsgClawback = "msg_clawback"
+
 var _ sdk.Msg = &MsgCreateVestingAccount{}
 
 var _ sdk.Msg = &MsgCreatePeriodicVestingAccount{}
 
+var _ sdk.Msg = &MsgCreateTrueVestingAccount{}
+
+var _ sdk.Msg = &MsgClawback{}
```
nit, group

JimLarson replied:
Done.

fedekunze commented on x/auth/vesting/types/period.go:
```diff
@@ -26,3 +27,257 @@ func (vp Periods) String() string {
+	time := startTime
+
+	for _, period := range periods {
+		x := readTime - time
+		if x < period.Length {
+			break
+		}
```
I feel this could have better name + comment

JimLarson replied:
I had copied this from `vesting_account.go`, so I made minimal changes for ease of review. But I'm happy to refactor. Done.

fedekunze commented on x/auth/vesting/types/period.go:
```diff
@@ -26,3 +27,257 @@ func (vp Periods) String() string {
+// events combined into a single event.
+// Returns new start time, new end time, and merged vesting events, relative to
+// the new start time.
+func DisjunctPeriods(startP, startQ int64, p, q []Period) (int64, int64, []Period) {
```
this code is really hard to read and thus not very maintainable. What is P, what is Q?

JimLarson replied:
Inputs `p` and `q` are the sequences of periods of the two schedules to merge. It's just notation, like inputs `a` and `b` to `coinsMin()` right above.

I've made it more consistent (but less compact) by renaming `p` and `q` to `periodsP` and `periodsQ` respectively.

fedekunze commented on x/auth/vesting/types/period.go:
```diff
@@ -26,3 +27,257 @@ func (vp Periods) String() string {
+		iQ++
+	}
+
+	for iP < lenP && iQ < lenQ {
+		nextP := timeP + p[iP].Length // next p event in absolute time
+		nextQ := timeQ + q[iQ].Length // next q event in absolute time
+		if nextP < nextQ {
+			consumeP(nextP)
+		} else if nextP > nextQ {
+			consumeQ(nextQ)
+		} else {
+			consumeBoth(nextP)
+		}
+	}
```
can you provide more comments about this block? What are the `iP` and `iQ`? what does consume do?

JimLarson replied:
The variables and helper functions are commented above. I've added a comment to describe what the loop is doing. I tried making the index variables more verbose, but the result didn't seem as legible, so I've kept them terse.

Ultimately, this is an inherently fussy little algorithm of simultaneous iteration over relative time periods, and there's limits to how far comments and identifier names can mitigate that fussiness.

fedekunze commented on x/staking/keeper/delegation.go:
```diff
@@ -626,6 +627,152 @@ func (k Keeper) Delegate(
+
+	// Check redelegation entry limits while we can still return early.
+	// Assume the worst case that we need to transfer all redelegation entries
+	toRedelegations := k.GetRedelegations(ctx, toAddr, math.MaxUint16)
```
why not iterate over the redelegations?

JimLarson replied:
Doing the batch for consistency with line 708 below, but since this is read-only I could change it if you feel strongly.

fedekunze replied:
I'd rely on the `Iterator` to avoid iterating twice `GetRedelegations` and `for` loop below

JimLarson replied:
Switched to an iterator. The existing redelegation iterator is a full table scan, so implemented a new one for a single delegator.

fedekunze commented on x/staking/keeper/delegation.go:
```diff
@@ -626,6 +627,152 @@ func (k Keeper) Delegate(
+	// keeping a liability for 25 shares and transferring one for 75 shares.
+	// Of course, the redelegations themselves can have multiple entries for
+	// different timestamps, so we're actually working at a finer granularity.
+	redelegations := k.GetRedelegations(ctx, fromAddr, math.MaxUint16)
```
ditto use iterator

JimLarson replied:
It looks like you break the contract if you modify the section of the store you're iterating over.  See comment on `Iterator()` method of `KVStore` in `store/types/store.go`.

fedekunze commented on x/distribution/keeper/delegation.go:
```diff
@@ -136,6 +136,16 @@ func (k Keeper) CalculateDelegationRewards(ctx sdk.Context, val stakingtypes.Val
+// SmartRewardAccount is an account with a post-reward processing function.
+// Such an account ignores a non-default withdrawal address, as this can
+// be implemented in the post-reward processing if desired.
+type SmartRewardAccount interface {
```
why not introduce Distribution Hooks instead of a new account?

JimLarson replied:
See below.

fedekunze commented on x/distribution/keeper/delegation.go:
```diff
@@ -166,11 +176,20 @@ func (k Keeper) withdrawDelegationRewards(ctx sdk.Context, val stakingtypes.Vali
 			return nil, err
 		}
+		if isSmart {
+			smartAcc.PostReward(ctx, coins, k.authKeeper, k.bankKeeper, k.stakingKeeper)
```
I think this should be an `AfterWithdrawRewards` hook.

JimLarson replied:
We also need to prevent the use of a different `withdrawAddr` when we're dealing with a vesting account with clawback. I could have another hook method, `AllowWithdrawAddr(acc authtypes.AccountI) bool` - although this is odd, since there are no other hook methods that return a value or affect execution of the calling module.

Vesting accounts themselves are precedent for having modules modify their behavior based on account type - in its interaction with the `x/bank` module. However, I'm not sure if that's a good precedent to follow.

So I could use the following odd hook interface.  I'm not sure if the resulting code is any simpler or more idiomatic:

```
type interface DistributionHook {
  AllowWithdrawAddr(acc authtypes.AccountI) bool
  AfterReward(ctx sdk.Context, acc authTypes.AccountI, reward sdk.Coins)
}
```

Or I could keep this as-is. Let me know which you'd prefer.

(BTW, why don't the hook-calling modules allow for an arbitrary number of registered hooks? It doesn't seem any harder.)

```
type StakingKeeper struct {
    ...
    hooks []types.StakingHooks
    ...
}

func (k Keeper) BeforeFoo(ctx sdk.Context, addr sdk.AccAddress, foo types.Whatever) {
    for _, hook := range k.hooks {
        hook.BeforeFoo(ctx, addr, foo)
    }
}
```

JimLarson replied:
Okay, so I implemented distribution hooks as I described. It was a little tricky getting the wiring correct to tie it all together, but now it's more general, and I think we'll have some other uses for it.

fedekunze commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -611,3 +589,412 @@ func marshalYaml(i interface{}) (interface{}, error) {
+	unbondings := sk.GetUnbondingDelegations(ctx, tva.GetAddress(), math.MaxUint16)
+	for _, unbonding := range unbondings {
+		for _, entry := range unbonding.Entries {
+			unbonded = unbonded.Add(entry.Balance)
```
shouldn't this be unbonding?

JimLarson replied:
Yes, it should, thanks. I'll expand the test coverage for this.

fedekunze commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -611,3 +589,412 @@ func marshalYaml(i interface{}) (interface{}, error) {
+		tokens := validator.TokensFromSharesTruncated(shares).RoundInt()
+		bonded = bonded.Add(tokens)
+	}
+	return
```
nit, the return values should be explicit

JimLarson replied:
Done.

fedekunze reviewed (commented):
This implementation has a lot of nice additions to the vesting logic. Unfortunately, it's really hard to follow through. I'd highly recommend using specific names for variables, adding more comments on top of the functions and inside them to complement the code and make it more maintainable.

As for the staking and distribution, I think they can be split into a separate PR and upstreamed to the SDK. I highly recommend using a distribution hook for rewards instead of a SmartAccount.

JimLarson commented:
@fedekunze My apologies for the lack of description. I'll be updating the vesting spec doc, but in the meantime please refer to https://github.com/Agoric/agoric-sdk/issues/4085

If you feel like an interactive discussion would be useful, please refer me to a discord channel, or feel free to schedule a video chat. I'm on Pacific time but would be happy to stay up late or get up early.

JimLarson committed:
fix: address review feedback - comments, naming, fixes

JimLarson commented:
@fedekunze PTAL, and thanks

I'll be online early-ish in the San Francisco timezone if you'd like to chat.

fedekunze commented:
Thanks @JimLarson, I sent Dean a Slack Connect invite so that we can talk about the PR there. I'll review it shortly
cc @zmanian

fedekunze commented on x/auth/vesting/msg_server.go:
```diff
@@ -160,5 +185,180 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+
+	acc := ak.GetAccount(ctx, addr)
+	if acc == nil {
+		return nil, fmt.Errorf("account %s does not exist", msg.Address)
```
use `sdkerrors` for msg server and Keeper errors

JimLarson replied:
Done, here and below.

fedekunze commented on x/auth/vesting/types/msgs_test.go:
```diff
@@ -79,7 +79,114 @@ func TestPeriodicVestingAccountMsg(t *testing.T) {
+	tp := okMsg.Type()
+	require.Equal(t, TypeMsgClawback, tp)
+	err := okMsg.ValidateBasic()
+	require.Nil(t, err)
```
use `require.NoError(t, err)` and `require.Error(t, err)` instead of `require.Nil(t, err)` and `require.NotNil(t, err)`

JimLarson replied:
Done, here and elsewhere.

fedekunze commented on x/auth/vesting/types/period.go:
```diff
@@ -26,3 +27,267 @@ func (vp Periods) String() string {
+}
+
+// coinsMin returns the minimum of its inputs for all denominations.
+func coinsMin(a, b sdk.Coins) sdk.Coins {
```
nit I think this should be a func of the Coins type and should be split and upstreamed to the SDK

`func (c Coins) Min(other Coins) Coins`

JimLarson replied:
Will do when I get a chance. `Coins` also needs an equality test that doesn't panic. I've got some work in progress for `agoric-sdk` that adds some utilities with tests for their algebraic properties: https://github.com/Agoric/agoric-sdk/blob/3991-standalone-loc/golang/cosmos/types/coin.go

fedekunze replied:
can you create an issue?

JimLarson replied:
Created https://github.com/cosmos/cosmos-sdk/issues/10995

Please assign me if you wish - it's not letting me assign myself.

fedekunze commented on x/auth/vesting/types/period.go:
```diff
@@ -26,3 +27,267 @@ func (vp Periods) String() string {
+	min := sdk.NewCoins()
+	for _, coinA := range a {
+		denom := coinA.Denom
+		bAmt := b.AmountOf(denom)
```
```suggestion
		bAmt := b.AmountOfNoValidation(denom)
```

JimLarson replied:
Done.

liamsi commented:
@fedekunze can you also add me and @musalbas to that shared slack if possible?

JimLarson committed:
fix: more review feedback

- Use proper sdkerrors where appropriate.
- Use correct test helpers for checking errors.
- Implement a new redelegation iterator and use it in read-only
  cases.

fedekunze commented on x/auth/vesting/client/cli/tx.go:
```diff
@@ -129,35 +164,95 @@ func NewMsgCreatePeriodicVestingAccountCmd() *cobra.Command {
+		
+		A periods file is a JSON object describing a sequence of unlocking or vesting events,
+		with a start time and an array of coins strings and durations relative to the start or previous event.
+		For example:
```
nit, use `Example` field?

JimLarson replied:
Done, here and above.

fedekunze commented on x/auth/vesting/msg_server.go:
```diff
@@ -118,35 +122,55 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
-	acc := types.NewPeriodicVestingAccount(baseAccount.(*authtypes.BaseAccount), totalCoins.Sort(), msg.StartTime, msg.VestingPeriods)
+	if acc != nil {
+		pva, ok := acc.(*types.PeriodicVestingAccount)
+		if !msg.Merge {
+			if ok {
+				return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s already exists; consider using --merge", msg.ToAddress)
+			}
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s already exists", msg.ToAddress)
+		}
+		if !ok {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrNotSupported, "account %s must be a periodic vesting account", msg.ToAddress)
+		}
```
can we use `switch` here instead?

JimLarson replied:
Done.

fedekunze commented on x/auth/vesting/msg_server.go:
```diff
@@ -160,5 +184,180 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+	if acc != nil {
+		var ok bool
+		va, ok = acc.(*types.TrueVestingAccount)
+		if !msg.Merge {
+			if ok {
+				return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s already exists; consider using --merge", msg.ToAddress)
+			}
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s already exists", msg.ToAddress)
+		}
+		if !ok {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrNotSupported, "account %s must be a true vesting account", msg.ToAddress)
+		}
+		if msg.FromAddress != va.FunderAddress {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s can only accept grants from account %s", msg.ToAddress, va.FunderAddress)
+		}
```
use `switch` for readibility

JimLarson replied:
Done.

fedekunze commented on x/auth/vesting/msg_server.go:
```diff
@@ -160,5 +184,180 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s can only accept grants from account %s", msg.ToAddress, va.FunderAddress)
+		}
+		newStart, newEnd, newLockupPeriods := types.DisjunctPeriods(va.StartTime, msg.GetStartTime(), va.LockupPeriods, msg.LockupPeriods)
+		newStartX, newEndX, newVestingPeriods := types.DisjunctPeriods(va.StartTime, msg.GetStartTime(),
```
what does  `X` mean here?

JimLarson replied:
Renamed for clarity. I really ought to switch to a sequence of (timestamp, amount) pairs instead of these relative time periods with external start/end - but not now.

fedekunze commented on x/auth/vesting/msg_server.go:
```diff
@@ -160,5 +184,180 @@ func (s msgServer) CreatePeriodicVestingAccount(goCtx context.Context, msg *type
 	return &types.MsgCreatePeriodicVestingAccountResponse{}, nil
+}
+
+func max64(a, b int64) int64 {
```
isn't this a dup from the function in types/?

JimLarson replied:
Yes, and IMHO it's a trivial enough function that it's fine to define locally where needed. E.g. the standard library defines it for `int` but doesn't bother to export it.

JimLarson committed:
refactor: implement general distribution hooks

...instead of magic account type.

JimLarson committed:
refactor: more review responses

Use cobra command example documentation.
Switch instead of chained/branching if-statements.
Some renaming.

JimLarson committed:
fix: use from flag to specify funder in clawback command

Unfortunately confusing semantics for this command,
but least surprising given command framework.

JimLarson committed:
test: cli tests for clawback vesting

JimLarson committed:
refactor: rename TrueVestingAccount and friends to ClawbackVestingAccount

JimLarson committed:
docs: update vesting spec for clawback accounts

JimLarson committed:
fix: correctly handle post-clawback residual lockup after slashing

JimLarson committed:
test: expand tests, test clawback

JimLarson committed:
refactor: remove 'combined_periods' fields and compute dynamically

JimLarson committed:
fix: clawback after slash fix, refactor test helper

JimLarson committed:
fix: factor out reward when computing reward vesting

JimLarson committed:
refactor: put additional grant merge logic in vesting accounts

JimLarson committed:
fix: reset delegation bookkeeping when adding new grants

Fixes issue https://github.com/Agoric/agoric-sdk/issues/4300

JimLarson committed:
fix: remember slashed amount when adding vesting grant

JimLarson committed:
test: corner case for add grant after full slash

Make sure we can merge with an empty vesting schedule.

JimLarson committed:
refactor: provide staking routines for bonded and unbonding balances

JimLarson committed:
test: improve coverage, fix some test bugs

JimLarson commented:
@michaelfig PTAL at the last few changes.

michaelfig reviewed (approved):
LGTM!
